web/src/channel.ts

```diff
@@ -151,41 +162,43 @@ function call(args) {
     }
 }
 
-export function get(options) {
+export function get(options: AjaxRequestHandlerOptions): JQuery.jqXHR<unknown> | undefined {
     const args = {type: "GET", dataType: "json", ...options};
     return call(args);
 }
 
-export function post(options) {
+export function post(options: AjaxRequestHandlerOptions): JQuery.jqXHR<unknown> | undefined {
     const args = {type: "POST", dataType: "json", ...options};
     return call(args);
 }
 
-export function put(options) {
+export function put(options: AjaxRequestHandlerOptions): JQuery.jqXHR<unknown> | undefined {
     const args = {type: "PUT", dataType: "json", ...options};
     return call(args);
 }
 
 // Not called exports.delete because delete is a reserved word in JS
-export function del(options) {
+export function del(options: AjaxRequestHandlerOptions): JQuery.jqXHR<unknown> | undefined {
     const args = {type: "DELETE", dataType: "json", ...options};
     return call(args);
 }
 
-export function patch(options) {
+export function patch(
+    options: AjaxRequestHandlerOptions & {data: FormData | Record<string, unknown>},
+): JQuery.jqXHR<unknown> | undefined {
     // Send a PATCH as a POST in order to work around QtWebkit
     // (Linux/Windows desktop app) not supporting PATCH body.
     if (options.processData === false) {
         // If we're submitting a FormData object, we need to add the
         // method this way
         options.data.append("method", "PATCH");
     } else {
-        options.data = {...options.data, method: "PATCH"};
+        options.data = {...(options.data as Record<string, unknown>), method: "PATCH"};
```

Comment:
Seems when `data` is `FormData`, we run `if` statement, while when `data` is `Record<string, unknown>`, we run `else` statement. Played around with the types of `option` a bit and I figured out if we write 
```
options: AjaxRequestHandlerOptions &({processData: false, data: FormData} | {processData?: true, data: Record<string, unknown>}) 
```
then we won't need to casting `options.data` to `Record<string, unknown>`, though the annotation is much messier.

Reply:
This does a little messier but also I tried this type definition and it seems it isn't working here not completely sure why though.

Reply:
Looks like this doesn't work:
```diff
diff --git a/web/src/channel.ts b/web/src/channel.ts
index 246aa4dd9e..20c405c5ad 100644
--- a/web/src/channel.ts
+++ b/web/src/channel.ts
@@ -188,7 +188,7 @@ export function del(options: AjaxRequestHandlerOptions): JQuery.jqXHR<unknown> |
 }
 
 export function patch(
-    options: AjaxRequestHandlerOptions & {data: FormData | Record<string, unknown>},
+    options: AjaxRequestHandlerOptions & ({processData: false, data: FormData} | {processData?: true, data: Record<string, unknown>}) 
 ): JQuery.jqXHR<unknown> | undefined {
     // Send a PATCH as a POST in order to work around QtWebkit
     // (Linux/Windows desktop app) not supporting PATCH body.
@@ -197,7 +197,8 @@ export function patch(
         // method this way
         options.data.append("method", "PATCH");
     } else {
-        options.data = {...(options.data as Record<string, unknown>), method: "PATCH"};
+        const data: (string & Record<string, unknown>) | (JQuery.PlainObject & Record<string, unknown>) = options.data;
+        options.data = {...data, method: "PATCH"}; // Spread types may only be created from object types.
     }
     return post(options);
 }
```

Reply:
But this works:
```
diff --git a/web/src/channel.ts b/web/src/channel.ts
index 246aa4dd9e..59c304e819 100644
--- a/web/src/channel.ts
+++ b/web/src/channel.ts
@@ -188,7 +188,7 @@ export function del(options: AjaxRequestHandlerOptions): JQuery.jqXHR<unknown> |
 }
 
 export function patch(
-    options: AjaxRequestHandlerOptions & {data: FormData | Record<string, unknown>},
+    options: AjaxRequestHandlerOptions & ({processData: false, data: FormData} | {processData?: true, data: Record<string, unknown>}) 
 ): JQuery.jqXHR<unknown> | undefined {
     // Send a PATCH as a POST in order to work around QtWebkit
     // (Linux/Windows desktop app) not supporting PATCH body.
@@ -197,7 +197,8 @@ export function patch(
         // method this way
         options.data.append("method", "PATCH");
     } else {
-        options.data = {...(options.data as Record<string, unknown>), method: "PATCH"};
+        const data: Record<string, unknown> = options.data;
+        options.data = {...data, method: "PATCH"}; // Spread types may only be created from object types.
     }
     return post(options);
 }
```

Reply:
This probably is simpler though:
```diff
diff --git a/web/src/channel.ts b/web/src/channel.ts
index 246aa4dd9e..87fe34f743 100644
--- a/web/src/channel.ts
+++ b/web/src/channel.ts
@@ -188,7 +188,7 @@ export function del(options: AjaxRequestHandlerOptions): JQuery.jqXHR<unknown> |
 }
 
 export function patch(
-    options: AjaxRequestHandlerOptions & {data: FormData | Record<string, unknown>},
+    options: AjaxRequestHandlerOptions & ({processData: false, data: FormData} | {processData?: true, data: Record<string, unknown>}) 
 ): JQuery.jqXHR<unknown> | undefined {
     // Send a PATCH as a POST in order to work around QtWebkit
     // (Linux/Windows desktop app) not supporting PATCH body.
@@ -197,7 +197,7 @@ export function patch(
         // method this way
         options.data.append("method", "PATCH");
     } else {
-        options.data = {...(options.data as Record<string, unknown>), method: "PATCH"};
+        options.data.method = "PATCH";
     }
     return post(options);
 }
```

Reply:
Yeah, I don't see any reason to not mutate `options` here and this works.

But can we maybe omit `data` field entirely from `AjaxRequestHandlerOptions` and take it's intersection with `({processData: false, data: FormData} | {processData?: true, data: Record<string, unknown>})`?

That way the `data` field will only have `FormData` and `Record<string, unknown>` as acceptable data type for `patch` request and I have checked that every patch request does indeed use only these two types. It makes sense to drop the `string` type entirely here?

Reply:
I have done this change in the latest changes, let me know if this approach works for you.

Reply:
Yeah this looks good!